src/libstore/s3-binary-cache-store.cc

```diff
@@ -183,6 +183,63 @@ S3Helper::FileTransferResult S3Helper::getObject(
 
     return res;
 }
+S3Helper::StreamResult S3Helper::getObjectStreamed(
+    const std::string & bucketName, const std::string & key, Sink & sink, size_t bufferSize)
+{
+    debug("streaming from 's3://%s/%s'...", bucketName, key);
+    auto now1 = std::chrono::steady_clock::now();
+    S3Helper::StreamResult res;
+
+    try {
+        size_t size = getObjectSize(bucketName, key);
```

Comment:
This adds a `HEAD` request for every `GET` request, which I imagine adds a substantial amount of latency, especially for `.narinfo` files (which are very small). Do we really need to get the size in advance? Maybe we can start fetching from offset 0 and the response header will tell us what the full size is.

Reply:
I'll give it a try. Let's hope the S3 HTTP API supports this.